### PR TITLE
New nn API

### PR DESCRIPTION
**Pull request by pabloferz**
This is a full rewrite of the `nn` submodule. The idea is to merge this in, then switch all the methods that use neural nets to use this instead and then to remove the previous submodule.

### CODE REVIEW

File: pysages/_nn/objectives.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    prod,
+    sum_squares,
+    unpack,
+)
+
+import jax.numpy as np
+import numpy as onp
+
+
+# Losses
```
Comment (InnocentBug) [inline pysages/_nn/objectives.py:22]: Are these all abstract classes, or why are they all defined here, but not implemented.
I am a little confused here.

File: pysages/_nn/objectives.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+class L2Regularization(Regularizer, NamedTuple):
+    coeff: Float
+
+
+class VarRegularization(Regularizer):
+    pass
+
+
+@dispatch
+def build_cost_function(model, loss: Loss, reg: Regularizer):
```
Comment (InnocentBug) [inline pysages/_nn/objectives.py:61]: this function overloading is a bit confusing to me.
Is it really the best way to have overloaded `build_cost_function` with different number of arguments.
Wouldn't it make sense to have a different name for the ones that take a model?

File: pysages/_nn/objectives.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    cost = build_cost_function(loss, reg)
```
Comment (InnocentBug) [inline pysages/_nn/objectives.py:62]: I don't understand, why this is generated here.
Does it have a side effects?
I don't see where `cost` would be used after.

File: pysages/_nn/objectives.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        prediction = model.apply(params, inputs).reshape(reference.shape)
+        e = np.asarray(prediction - reference, dtype = np.float32).flatten()
+        ps, _ = unpack(params)
+        return cost(e, ps)
+
+    return objective
+
+
+@dispatch
+def build_cost_function(model, loss: Sobolev1Loss, reg: Regularizer):
```
Comment (InnocentBug) [inline pysages/_nn/objectives.py:74]: I guess, for all of them. It wouldn't hurt to have a doc string, that explains what the model is and how it is implemented here.

**Review by InnocentBug — CHANGES_REQUESTED** [review]
Some of the implementations confuse me.
But maybe that is my limited understanding of how dispatch is supposed to be used.
Or some documentation could help.

I also think, that an example that uses the code would make it easier for users to get into the code.

**Review by InnocentBug — APPROVED** [review]
I can't really test it. But it looks good.

**Comment by pabloferz** [comment]
I'll add a README and tests in a follow up PR.

**Pull request merged**